topydo/lib/ChangeSet.py

```diff
@@ -0,0 +1,94 @@
+# Topydo - A todo.txt client written in Python.
+# Copyright (C) 2014 - 2015 Bram Schoenmakers <me@bramschoenmakers.nl>
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+""" This module serves for managing todo and archive changesets. """
+
+from topydo.lib.Config import config
+from hashlib import sha1
+import shelve
+from copy import deepcopy
+
+class ChangeSet(object):
+    def __init__(self, p_todolist=None, p_archive=None):
+        self.todolist = deepcopy(p_todolist)
+        self.todolist_hash = self._hash_todolist(p_todolist)
+        self.archive = deepcopy(p_archive)
+        self.archive_hash = self._hash_todolist(p_archive)
+        self.shelve_file = shelve.open(config().backup_file())
+
+    def add_archive(self, p_archive):
+        """ Sets deep copy of p_archive as archive attribute and calculate archive_hash. """
+        self.archive = deepcopy(p_archive)
+        self.archive_hash = self._hash_todolist(p_archive)
+
+    def add_todolist(self, p_todolist):
+        """ Sets deep copy of p_todolist as todolist attribute and calculate todolist_hash. """
+        self.todolist = deepcopy(p_todolist)
+        self.todolist_hash = self._hash_todolist(p_todolist)
+
+    def save(self, p_todolist):
+        """
+        Saves a tuple with archive, todolist and their hashes into the
+        backup file. Tuple is indexed in backup file  with hash calculated from
+        p_todolist.
+        """
+        current_hash = self._hash_todolist(p_todolist)
+        self.shelve_file[current_hash] = (self.todolist, self.todolist_hash, self.archive, self.archive_hash)
+
+        self.close()
+
+    def delete(self, p_todolist_hash):
```

Comment:
I had to revert it in ad2fe7443c56a0c1caa53c632109c54d2a18a101 because this way it is much more simple to remove Changeset after it has been used in `undo` (feccf937b39a59428c80c493d760a87db6602e51). Otherwise we would have to make another `deepcopy` of `TodoList`.